RFC for doc_cfg, doc_cfg_auto, doc_cfg_hide and doc_cfg_show features

GuillaumeGomez opened the pull request:
[tracking issue](https://github.com/rust-lang/rust/issues/43781)

cc @rust-lang/rustdoc

[Rendered](https://github.com/GuillaumeGomez/rfcs/blob/doc-cfg/text/000-rustdoc-cfgs-handling.md)

Nemo157 commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+### `#[doc(cfg(...))]`
+
+This attribute provides a standardized format to document conditionally available items. Example:
```
I would describe this primarily as being about _overriding_ `#[cfg]` for docs, with the intent of having `doc(auto_cfg)` being the default (and maybe swapping their order in the explanation).

Nemo157 commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+### `#[doc(auto_cfg)]`/`#[doc(no_auto_cfg)]`
+
+By default, `#[doc(auto_cfg)]` is enabled at the crate-level. When it's enabled, Rustdoc will automatically display `cfg(...)` compatibility information as-if the same `#[doc(cfg(...))]` had been specified.
```
Do we want to change this default in existing code, or maybe this should be edition tied?

GuillaumeGomez replied:
As far as I can see, it doesn't bring downside to have more information displayed by default, so I think it shouldn't be tied to an edition.

Nemo157 commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+In some situations, the detailed conditional compilation rules used to implement the feature might not serve as good documentation (for example, the list of supported platforms might be very long, and it might be better to document them in one place). To turn it off, add the `#[doc(no_auto_cfg)]` attribute.
+
+Both `#[doc(auto_cfg)]` and `#[doc(no_auto_cfg)]` attributes impact all there descendants. You can then enable/disable them by using the opposite attribute on a given item. They can be used as follows:
```
The current implementation only allows these attributes at the crate-level, has there been a push to support this arbitrary nesting from somewhere?

Nemo157 replied:
(Well, actually the attributes don't exist, `feature(doc_auto_cfg)` _implies_ default-active `doc(auto_cfg)`, but IIRC previous discussions about how to expose a stable toggle had only considered adding in the attribute toggling it at the crate-level).

GuillaumeGomez replied:
> The current implementation only allows these attributes at the crate-level, has there been a push to support this arbitrary nesting from somewhere?

No, it's a change brought by this RFC: users will be able to use them on items directly if they want to have better control where they want it to be enabled or not.

Just like lints, I think it's better to give better control over this. For example if someone wants to completely override doc cfg attributes on a given item, they will be able to disable it for this item completely and then add what they want using `doc(cfg())`. For this I'm thinking mostly about if you have "parent" features with a lot of children and for some reasons you have a complex `cfg` to handle it and want to simplify its display when rendered in documentation.

Nemo157 replied:
`#[doc(cfg(...))]` already overrides `#[cfg(...)]` on a single item, you don't need to disable the auto-cfg for that.

The use case where you would want to disable `doc(auto_cfg)` on a sub-tree seems to be if you have a lot of `cfg` on sub-items of that tree that you don't want rendered at all (but these `cfg` _aren't_ ones you want to globally `#[cfg_hide(...)]`), and so by disabling it you get to avoid putting `#[doc(cfg())]` on each sub-item individually. But that seems like an unlikely situation to me.

Similarly wanting to `#[cfg_hide(...)]` some `cfg` within just a sub-tree seems very unlikely,

Making these attributes work within the module tree brings in extra complexity by requiring the inverses to exist. If they're restricted to applying only at the crate root we only need two new attributes: `#[doc(no_auto_cfg)]` (assuming the default is active) and `#[doc(cfg_hide(...))]`.

GuillaumeGomez replied:
Fair point. I'll limit `doc(auto_cfg)` as a crate-level attribute.

For `cfg_hide` I'm less convinced as you might want to hide some cfgs only for some modules/items (this is very hypothetical though). As for the complexity, it seems pretty low in a visitor: we add new entries before iterating to the children and remove them once we leave the current item (but that's just an implementation detail, not sure it's something that we should take into account here).

Anyway, if no one seems to be interested into having `cfg_show`/`cfg_hide` to be used on a specific module/item, then I will turn them into crate-level attribute (well, `cfg_show` would be useless in this case so it'll be simply removed).

GuillaumeGomez replied:
I just thought about a use case where having `doc(auto_cfg)` not only as a crate-level attribute would be a nice improvement! In sysinfo, I disabled the `doc_auto_cfg` because every item is available, whatever the OS (even unsupported ones). Except that I recently splitted the components to enable them through features. So I can either do:

```rust
#![doc(auto_cfg = true)]
#![doc(cfg_hide(/* list of all supported OSes */)]
```

Or instead disable `auto_cfg` at the crate level and enable it on the components module directly without needing to worry about hiding `cfg`s since the OS information handling is one level upper.

So with this in mind, I think we should lift the crate-level only restriction on `auto_cfg`. What do you think?

Nemo157 commented on text/000-rustdoc-cfgs-handling.md:
````diff
@@ -0,0 +1,283 @@
+#![doc(cfg_hide(doc))]
+```
+
+Or directly on a given item/module as it covers any of the item's descendants:
````
Again the current implementation only supports being set at the crate-level, are there usecases that want to be able to limit it to certain sub-trees?

GuillaumeGomez replied:
The one case I'm thinking about is to reduce complexity of rendered doc cfg as described in [this comment](https://github.com/rust-lang/rfcs/pull/3631#discussion_r1597427115).

Nemo157 commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+Then, the `doc` cfg will never be displayed into the documentation.
+
+Rustdoc currently hides `doc` and `doctest` attributes by default and reserves the right to change the list of "hidden by default" attributes.
```
It would be good to not use default-hidden predicates in the examples above then (another good common example could be `fuzzing`).

GuillaumeGomez replied:
👍

kennytm commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+![This is supported on feature="futures-io" only.](https://user-images.githubusercontent.com/81079/89731116-d7b7ce00-da44-11ea-87c6-022d192d6eca.png)
+
+This attribute has the same syntax as conditional compilation, but it only causes documentation to be added. This means `#[doc(cfg(false))` will not cause your docs to be hidden, even though `#[cfg(false)]` does do that.
```
```suggestion
This attribute has the same syntax as conditional compilation, but it only causes documentation to be added. This means `#[doc(cfg(false))]` will not cause your docs to be hidden, even though `#[cfg(false)]` does do that.
```

btw I do wish we have `#[cfg(true)]` and `#[cfg(false)]`, lol, but it is currently a compile error

```rust
error: `cfg` predicate key cannot be a literal
 --> src/lib.rs:1:7
  |
1 | #[cfg(false)]
  |       ^^^^^
```

GuillaumeGomez replied:
We have `cfg(any())` and `cfg(all())` that kinda does that. ;)

Thanks for the typo fix!

clarfonthey commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+Of course, the above example is equivalent to "available on **Windows** only."
+
+Making this actually work all the time is equivalent to a [boolean satisfiability] check, coliquially called a "SAT problem," and can take exponential time.
```
This is… actually not true. The format you'd expect is probably akin to [dysjunctive normal form](https://en.wikipedia.org/wiki/Disjunctive_normal_form) or [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form) and converting to either of these is a polynomial-time algorithm.

I wouldn't expect this to be a requirement to stabilise, but, just wanted to correct the statement that this would take exponential time. Simplifying expressions actually is polynomial time; the exponential-time algorithm is figuring out the cases where the `cfg` could actually be applied, which is beyond the scope of rustdoc.

clarfonthey replied:
It's a very easy mistake to make! Unfortunately it's not a common part of maths or even computer science education, but it is very useful for simplifying boolean expressions.

GuillaumeGomez replied:
Thanks for the explanations, it's very welcome!

GuillaumeGomez replied:
I removed the technical part as I think it's out of context here. The linked issue remains so the discussion can continue there. :)

clarfonthey commented:
I'm in favour of this but I would recommend clarifying the `#[cfg(any(doc, ...))]` trick in the guided level explanation a bit further, since folks reading it might be misled into thinking that rustdoc will just ignore conditional compilation entirely to report things, which isn't true; it just shows the conditions when they're met.

jhpratt commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+    These attributes suppress or un-suppress the `auto_cfg` behavior for a particular configuration predicate.
+
+    For example, `#[doc(cfg_hide(windows))]` shall be used in newer versions of the [`windows` crate] to prevent the "this is supported on **windows** only" tag from being shown on every single item.
```
```suggestion
    For example, `#[doc(cfg_hide(windows))]` could be used in newer versions of the [`windows` crate] to prevent the "this is supported on **windows** only" tag from being shown on every single item.
```

"Shall" would imply that they're being forced to do so, which isn't the case.

jhpratt reviewed (commented):
Was any thought given to what I mentioned [in the draft RFC](https://github.com/GuillaumeGomez/rfcs/pull/2#issuecomment-1441578687)? Correctly annotating re-exports from a different crate is something that would be very useful for façade  crates and similar, and is the sole blocker for refactoring `time` to be split along more sensible lines. That would reduce code duplication a significant amount.

Manishearth commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+### `#[doc(cfg_hide(...))]`
+
+This attribute is used to prevent some `cfg` to be generated in the visual markers. So in the previous example:
```
issue: please be clear that this and the other attribute only apply to `auto_cfg`

GuillaumeGomez replied:
Indeed, thanks!

Manishearth commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+### Re-exports and inlining
+
+`cfg` attributes of a re-export are never merged the re-exported item(s). If `#[doc(inline)]` attribute is used on a re-export, the `cfg` of the re-exported item will be merged with the re-export's.
```
issue: sentence two here contradicts the "never" in sentence one.

may also be worth expanding on "merged"

GuillaumeGomez replied:
Done!

Manishearth commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+    When this is turned on, `#[cfg]` attributes are shown in documentation just like `#[doc(cfg)]` attributes are.
+
+  * `#[doc(cfg_hide(...))]` / `#[doc(cfg_show(...))]`
```
issue: so i've never been happy with the sheer number of different doc options this feature introduces: the names are kind of all over the place and hard to predict.

I think `#[doc(cfg(..))]` is fine, but it would be nice if the rest all derived from each other.

Perhaps `#[doc(auto_cfg(disable))]`, `#[doc(auto_cfg(show(...), hide(...))]` etc? Ideally we only introduced _one_ new doc attribute but I'm not sure if that's easy since `cfg` can accept anything. Potentially could use the distinction between `cfg(..)` and `cfg = ...` here. Thoughts?

clarfonthey replied:
I personally wouldn't like using `cfg = ...` since it would just make the distinction more confusing. I'm not really sure there's a massive benefit to having `doc(auto_cfg(show(...)))` versus `doc(cfg_show(...))` since, while it is literally combining two attributes into one, it just feels like unnecessary extra typing to me.

That said, perhaps there would be a case for having `doc(auto_cfg(...))` for `doc(cfg_show(...))` and `doc(auto_cfg(not(...)))` for `doc(cfg_hide(...))`, since it would fit in with existing `cfg` and be a bit more discoverable. Perhaps `doc(auto_cfg(all))` and `doc(auto_cfg(none))` could be allowed as aliases for enabling/disabling all options.

Still has some issues, but, if we were to change the existing attributes, I'd prefer something like that over just combining the `show` and `hide` into a single `auto_cfg` attribute.

Manishearth replied:
@clarfonthey point taken on subtle syntax, I didn't really want to propose that anyway which is why I proposed it coming out of `auto_cfg`

To me the main distinction between `cfg_show` and `auto_cfg(show)` is the latter is easier to remember: there are many potential names for `cfg_show` and I won't remember which one it is. Especially since `auto_cfg` has `cfg` as the _second_ word. It's inconsistent and confusing.

If everything trees out of a single attribute it's easier to remember IMO, and not confusing.

Manishearth replied:
And yes I'm happy with experimentation with `auto_cfg`, I didn't want to propose one particular solution there just want it to be something straightforward, intuitive, and easy to remember.

clarfonthey replied:
I'm not exactly convinced that `auto_cfg(show(...))` is easier to remember compared to `cfg_show(...)`, and in both cases, IDEs can offer suggestions after typing `doc(cfg` to let you know what's available.

My main point here is that having "fewer attributes" isn't really a principle we should be guided by, since what matters most is the user writing the code, not the compiler parsing it. To a user, `auto_cfg(show(...))` and `auto_cfg(hide(...))` are still two attributes; they just require extra parentheses compared to `cfg_show(...)` and `cfg_hide(...)`.

Manishearth replied:
I do not think "everyone is using an IDE" is either correct or a driving design principle we should have.

My argument here is entirely from the user perspective. I find these names confusing and hard to remember, and a logical grouping would simplify it. I do not care one whit how easy or hard it is for the compiler to parse.

Manishearth replied:
I'm also open to that though I still prefer having fewer attributes

clarfonthey replied:
I agree with the notion that an IDE should not be assumed; that was tangential to the point that adding extra parentheses doesn't really affect the mental model of how to do things here, IMHO. Show and hide are a pair, whether they are rested inside an "auto" or not.

Manishearth replied:
Right, but they _do_ apply only to autocfg, it's part of what makes them confusing.

Nemo157 replied:
Another option could be `#[doc(auto_cfg_hide(...))]` then.

Though, I'm now wondering whether there's a way to change the semantics slightly so that `cfg_hide` can take over the role of `no_auto_cfg` too. Instead of disabling `auto_cfg` completely just tell it to hide _everything_ so it doesn't actually do anything. (Combined with my discussion above of restricting this to the crate-root that means there would only be two new attributes required in this RFC: `#[doc(cfg(...))]` anywhere and `#![doc(cfg_hide(...))]` at the root).

GuillaumeGomez replied:
I don't understand what you suggested @Nemo157. Do you mean that if no argument is provided to `cfg_hide`, it means that we hide everything?

So to summarize the discussion, it seems that we tend to for `show` and `hide`:

```rust
#[doc(cfg_show(test))]
#[doc(cfg_hide(test))]
```

As for enabling/disabling the feature, what do you think about:

```rust
#[doc(cfg_auto(enable)]
#[doc(cfg_auto(disable)]
```

?

clarfonthey replied:
My understanding of the issue, after completely misunderstanding for a while (apologies, @Manishearth) is that the `cfg_show` and `cfg_hide` attributes make no reference to `auto_cfg` and so it's not clear that they control *automatically* showing and hiding `cfg` attributes. Manish' proposal of `auto_cfg(show(...))` and `auto_cfg(hide(...))` is one way of accomplishing this, but `auto_cfg_show` and `auto_cfg_hide` attributes could also accomplish this.

Nemo157 replied:
> I don't understand what you suggested @Nemo157. Do you mean that if no argument is provided to `cfg_hide`, it means that we hide everything?

I don't have a concrete syntax proposal, I just wanted to seed the idea in case someone else has thoughts on how it could be accomplished. Being able to reduce this feature to introducing a total of two attributes instead of five I think is a helpful simplification.

Manishearth replied:
@clarfonthey yes, thanks. my two main complaints are that the split of attributes starting and ending with cfg is confusing (and the potential for them to do either makes it harder to remember _even if_ we pick consistent names), and the second one is the one you describe: it's not clear that show/hide is about auto stuff.

GuillaumeGomez commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+### `#[doc(cfg_show(...))]`
+
+This attribute does the opposite of `#[doc(cfg_hide(...))]`: if you used `#[doc(cfg_hide(...))]` and want to revert its effect on an item and its descendants, you can use `#[doc(cfg_show(...))]`:
```
Good catch! In this case it should emit an error if you show/hide a doc cfg on the same item.

GuillaumeGomez commented:
Applied first round of review comments. Thanks everyone!

clarfonthey commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,281 @@
+
+### `#[doc(auto_cfg(enable))]`/`#[doc(auto_cfg(disable))]`
+
+This is a crate-level attribute. By default, `#[doc(auto_cfg)]` is enabled at the crate-level. When it's enabled, Rustdoc will automatically display `cfg(...)` compatibility information as-if the same `#[doc(cfg(...))]` had been specified.
```
Is there a reason why this has to be a crate-level attribute? I don't see why it couldn't be applied on a per-module basis, for example; I imagine this is mostly a limitation of the existing implementation. In that case, it might be worthwhile ensuring that uses outside the crate level are linted with a warning that says they currently have no use, but might in the future, similar to how the `coverage` attribute is used.

GuillaumeGomez replied:
You can see the discussion about why it's crate-level only [here](https://github.com/rust-lang/rfcs/pull/3631#discussion_r1595617210).

fmease replied:
We could add it to *Future Possibilities*.

GuillaumeGomez replied:
Very good point!

clarfonthey commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,281 @@
+The end goal being to provide this information automatically so that the documentation maintenance cost won't increase.
+
+
+# Guide-level explanation
```
I mentioned this as a regular comment, but I'll move it to a review so it gets noticed: I think that the guide-level explanation should clarify the `cfg(any(doc, ...))` trick for making `auto_cfg` work properly, since it's not immediately clear otherwise. Basically, conditional compilation is still applied before rustdoc runs, meaning you have to ensure that rustdoc can see code to document it. Then, because `auto_cfg` by default ignores `cfg(doc)`, only the other `cfg`s are documented.

GuillaumeGomez replied:
It was not ignored. You can see the RFC mentioning this [here](https://github.com/rust-lang/rfcs/pull/3631/files#diff-2a397ac636038db2e30773d6b828f499f05e0b511b59da0522ac75e2a9868001R16). If it's not clear enough, how would you word it?

kpreid commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+
+This RFC proposes to add the following attributes:
+
+  * `#![doc(auto_cfg(enable))]`/`#[doc(auto_cfg(disable))]`
```
This is totally a bikeshed comment, so please feel free to ignore it, but I think it would be better if the syntax for this was:

* `#![doc(auto_cfg = true)]`
* `#![doc(auto_cfg = false)]`

This introduces fewer novel names, hints that there is a single boolean state which is being “shadowed”, and is more like the style of optional-feature-enabling used in `Cargo.toml`, whereas `foo(enable)`/`foo(disable)` is not like any precedent I can think of in Rust syntax.

GuillaumeGomez replied:
Very good point!

madsmtm commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,283 @@
+[cfg attribute]: https://doc.rust-lang.org/reference/conditional-compilation.html
+[`windows` crate]: https://docs.rs/windows/latest/windows/
+
+All of these attributes can be added to a module or to the crate root, and they will be inherited by the child items unless another attribute overrides it (except that `doc(cfg)` cannot be added to the crate root). This is why "opposite" attributes like `cfg_hide` and `cfg_show` are provided: they allow a child item to override its parent.
```
> except that `doc(cfg)` cannot be added to the crate root

Huh, why not? Both `#![doc(cfg(...))]` and `#![cfg(...)]` currently seem to work as I'd expect?

GuillaumeGomez replied:
Good catch!

GuillaumeGomez commented:
Updated the syntax for `auto_cfg` to instead accept `true` or `false`.

Removed restriction on `doc(cfg())` which was marked as not being able to be used as a crate-level attribute as it currently works using it as such, so doesn't seem there is a good enough reason to change this behaviour.

petrochenkov commented:
A couple more questions.

- Is `#[doc(cfg)]` currently allowed on an item without `#[cfg]`? What are the semantics?
  - I'd expect this to be allowed making `cfg_show` redundant as in the previous comment.
- Is it possible to show a cfg on an item, but hide it on all children of that item? (Without putting attributes on all those children individually.)

GuillaumeGomez commented:
> I have re-read the reference part again, but I still don't understand why there are so many attributes.
> 
>     * Why is `#![doc(auto_cfg = true)]` necessary if it is the default?

Because future possibilities include to make this attribute work not only at crate-level but also on any item. And based on [this discussion](https://github.com/rust-lang/rfcs/pull/3631#discussion_r1614694568), I think it'll part of this RFC.

>     * Why is `#![doc(auto_cfg = false)]` necessary if it is equivalent to `#![doc(cfg_hide(any()))]`?

`cfg_hide` doesn't work like `cfg`: it only accepts "items". However, the RFC is very unclear about it, so I'll add it in the RFC.

>       * If it is not equivalent, then why it needs to do something different? It would be better if they were equivalent and `auto_cfg` didn't need to exist.

I have to admit I didn't think about `cfg_show` and `cfg_hide` to work this way at all. I'm not sure it's a good idea though. For example, how should we handle this case:

```rust
cfg_hide(any(windows, not(unix)))
```

How do we interpret the `not(unix)`? As a whole or should it invert the "hide unix"? So for simplicity, that's why `not`, `any` and `all` are not supported in `cfg_hide` and `cfg_show` and why we have `cfg_auto`.

>     * Why is `#![doc(cfg_show(...))]` necessary if it is equivalent to `#![doc(cfg(...))]`?

It is not. As mentioned in the RFC:

> It only applies to `#[doc(auto_cfg = true)]`, not to `#[doc(cfg(...))]`.

>       * If it is not equivalent, then why it needs to do something different? It would be better if they were equivalent and `cfg_show` didn't need to exist.

Because they both fill different role. You might want to add your own "doc cfg" information to either override what `auto_doc_cfg` generates or add your own information.


> A couple more questions.
> 
>     * Is `#[doc(cfg)]` currently allowed on an item without `#[cfg]`? What are the semantics?

The RFC says:

> This attribute provides a standardized format to override `#[cfg()]` attributes to document conditionally available items.
> ...
> This attribute has the same syntax as conditional compilation

So the semantics are covered I think. The RFC didn't make it clear it works with and without `cfg` on the item though, so I'll add a clarification.

>       * I'd expect this to be allowed making `cfg_show` redundant as in the previous comment.
> 
>     * Is it possible to show a cfg on an item, but hide it on all children of that item? (Without putting attributes on all those children individually.)

I think I see why you mention it. I suppose you talk a case like:

```
#[cfg(windows)]
mod module {
    #![doc(cfg_hide(windows))]

    fn item() {}
}
```

In this case, since the `doc(cfg_hide(windows))` is also an attribute of `module`, it should impact `module`. I think we should say that inner `doc(cfg_hide)` and `doc(cfg_show)` attributes don't impact the item if they are inner attributes.

GuillaumeGomez commented:
So only remains the question:

```rust
#[cfg(windows)]
mod module {
    #![doc(cfg_hide(windows))]

    fn item() {}
}
```

How to allow `cfg(windows)` to be displayed on `module` while having it hidden on its children? I think what the RFC offer is good enough:

```rust
#[cfg(windows)]
#[doc(cfg(windows)] // Will be displayed all the time
#[doc(cfg_hide(windows))] // Hides `windows` from `doc(auto_cfg)` rendering on `module` and its children
pub mod module {
    #[cfg(windows)] // Not displayed because of `cfg_hide(windows)`!
    pub fn item() {}
}
```

jhpratt commented:
Just wanted to make sure that my comment above (https://github.com/rust-lang/rfcs/pull/3631#pullrequestreview-2048876838) wasn't missed, as there has been no response or acknowledgement :slightly_smiling_face:

GuillaumeGomez commented:
I don't understand your comments. Do you have an example of what you would want to do?

jhpratt commented:
For an overly simplified case, see https://github.com/jhpratt/doc-cfg-demo

Running `cargo +nightly doc --all-features` on `alpha` will **incorrectly** say that `alpha::S::foo` is "Available on crate feature `foo-method` only." I hadn't checked this in a while, but previously it would have no annotation whatsoever. Ideally, the feature resolver would be utilized to show the correct annotation — it relies on the feature `beta-foo-method`.

This situation is quite common with façade crates and is the sole reason I have not migrated large amounts of code from `time` to `time-core`. Were I to do this, I could remove ~1000 LOC (off the top of my head) from `time-macros` that is currently duplicated.

GuillaumeGomez commented:
Okay so if I understand correctly what you say, reexported items should only show `cfg`s from the reexport's point of view (so to speak)? That sounds reasonable. Once you confirm it, I'll update the RFC.

jhpratt commented:
Effectively, yes. My point of view is "what does a consumer of the library care about". That would mean showing them what feature _they_ need to enable such that the item is present, taking into account feature transitivity, etc.

GuillaumeGomez commented:
I think with this all questions/comments were answered/resolved?

jhpratt commented:
As long as we agree on intended behavior, and it seems we do, then I'm satisfied. As I'd said above, the behavior had changed since I originally checked; it previously had _no_ annotation.

Feel free to use that example as a rustdoc test whenever that gets fixed, by the way. No credit necessary.

GuillaumeGomez commented:
I definitely plan to. :3

GuillaumeGomez commented:
I think it's time to start the FCP then.

@rfcbot fcp merge

madsmtm commented:
Friendly ping @Manishearth, @Nemo157, @aDotInTheVoid, @camelid, @jsha, this is blocking progress on stabilizing `#[doc(cfg(...))]` etc.

Manishearth commented:
I don't think my old concern that `auto_cfg` and `cfg_show` are confusing since they are new terms using cfg in different spots has been addressed.

I'm not going to block the RFC on that, but I don't want to approve it myself. If everyone else is fine with it that's fine, but I would like to see if the overall api surface can be made easier to remember.

workingjubilee commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,305 @@
+# Summary
+[summary]: #summary
+
+This RFC aims at providing rustdoc users the possibility to add visual markers to the rendered documentation to know under which conditions an item is available (currently possible through the following unstable features: `doc_cfg`, `doc_auto_cfg` and `doc_cfg_hide`).
+
+It does not aim to allow having a same item with different `cfg`s to appear more than once in the generated documentation.
+
+It does not aim to document items which are *inactive* under the current configuration (i.e., “`cfg`'ed out”).
```
...I'm slightly confused by this preamble, honestly, because I would figure the most important time to know when something is available is when it is, in fact, not available... when it has been "cfg'd out" but can be enabled by other means.

fmease replied:
You are right, this sentence doesn't tell the full story and should be improved (however, it's technically correct). I'm pretty sure the RFC does elaborate on it further down in the document (Ctrl+F `doc` (Match Case, Match Full Words)).

It implicitly refers to https://github.com/rust-lang/rust/issues/1998.

As you can probably guess, rustdoc doesn't and won't magically recompile the to-be-document crate under all possible configurations and combinations thereof to gather the information necessary to display the "portability info box".

So for the following crate, `function` won't show up in the generated docs unless you actually passed `--cfg special` to `rustdoc`:

```rs
#![feature(doc_auto_cfg)] // (A)
// #![feature(doc_cfg)] // (B)

// #[doc(cfg(special))] (B)
#[cfg(special)]
pub fn function() {}
```

Therefore, the common and offical workaround is the use of semi-special cfg (spec) `doc`:

```rs
#![feature(doc_auto_cfg)] // (A)
// #![feature(doc_cfg)] // (B)

// #[doc(cfg(special))] (B)
#[cfg(doc, special)]
pub fn function() {}
```

This gets more hairy if there are "multiple" functions called `function` (where the cfg specs differ for obvious reasons), one needs to choose with of the two functions to annotate with `doc`.

Of course, this doesn't scale to more complicated cases. E.g., if the parameter/return types differ by cfg spec. So, yeah. The general problem is not solved. That's what I meant with this sentence you highlighted.

workingjubilee replied:
I mean, I've certainly written documentation that needs that invocation (or similar) to work properly. I guess I'm just feeling a bit perplexed that there is seemingly quite a lot of new concepts in this RFC and that's left unsolved. I would expect an RFC that has the sense of adding a Lot Of Stuff to also have an answer to that critical missing piece, because it's a right pain in the ass! ...or I would expect it to be smaller and more tightly scoped.

I realize that's a Pure Vibes thing, but I do think it can be a bit hard to evaluate a solution... and thus people would come to consensus slower... if it feels incorrectly scoped.

jyn514 replied:
> rustdoc doesn't and won't magically recompile the to-be-documented crate under all possible cfgs and combinations thereof

i mean, rustc now tracks enough information that it doesn’t have to
https://github.com/rust-lang/rust/pull/109005

fmease replied:
> i mean, rustc now tracks enough information that it doesn’t have to

Well, rustc only stores the *name* of cfg'ed out items but rustdoc would need access to so much more (struct fields, function signatures, ...). See [`StrippedCfgItem`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/expand/struct.StrippedCfgItem.html).

GuillaumeGomez replied:
There actually wasn't an explanation about this in the RFC so I added one in the `Unresolved questions` section.

GuillaumeGomez commented:
I applied suggestions from @Manishearth about the `hide`/`show` attributes: since they're only applied on the `auto_cfg` feature cfgs, it makes more sense to embed them in the `auto_cfg` attribute (imo).

I also added an explanation about why `cfg`ed out items are not included by rustdoc.

Manishearth commented:
@GuillaumeGomez nice! Should the syntax for `auto_cfg` then be `doc(auto_cfg)` and `doc(auto_cfg = false)` (with `= true` being accepted but not the primary documented syntax)?

GuillaumeGomez commented:
You mean for `doc(auto_cfg)` to be equivalent to `doc(auto_cfg = true)`, making the `= true` optional?

GuillaumeGomez commented:
Not a super big fan of the implicit but it also kinda makes sense. Does it also include `doc(auto_cfg(hide(...)))`? Following the logic "if `auto_cfg` without `=` is used, it's enabled", it'd mean that `auto_cfg(hide)` would also enable `auto_cfg`. I'm fine with it but just want to confirm with you before updating the RFC.

Manishearth commented:
Yeah, `hide` doesn't make any sense without `auto_cfg`.

I mostly don't want to mix the `=` and the `()` syntax too much, and this way most normal users only use the `()` syntax

Manishearth commented:
At the very least, `doc(auto_cfg)` should work. My whole thrust here is to make the syntax relatively uniform and memorable.

GuillaumeGomez commented:
Ok, updating the RFC to reflect it then. 👍

GuillaumeGomez commented:
Added the new syntax (allowing to have no argument in `auto_cfg`) and also the new behaviour for `auto_cfg(hide(...))` and `auto_cfg(show(...))`.

Thanks for the ideas!

camelid commented on text/000-rustdoc-cfgs-handling.md:
````diff
@@ -0,0 +1,355 @@
+
+[Future versions of rustdoc][boolean simplification] may simplify this display down to "available on **Windows** only."
+
+## `doc(cfg)` and `doc(auto_cfg)` used on a same item
+
+If `doc(cfg)` is used on an item while `doc(auto_cfg)` is enabled, both will be merged:
+
+```rust
+// `doc(auto_cfg)` is enabled here
+#[doc(cfg(feature = "futures-io"))]
+#[cfg(feature = "something")]
+pub mod futures {}
+```
+
+In this case, it'll display that both `feature = "futures-io"` and `feature = "something"` are required to use this module. Just like lints, users will get better control over this feature. For example if someone wants to completely override doc cfg attributes on a given item, they will be able to disable it for this item completely and then add what they want using `doc(cfg)`. An example for thi ...
````
cc @rust-lang/rustdoc — Guillaume and I were discussing how this interaction should be handled because it wasn't originally explained in detail in the RFC text. This may not match your previous understanding though.

camelid replied:
It now overrides instead of merging, which is similar to what the text originally implied.

GuillaumeGomez replied:
I think it is what we agreed upon and it makes more sense in any case since `doc(cfg)` overrides `cfg`, it also implies that it overrides `doc(auto_cfg)`. But at least now it is written in the RFC. Thanks for that!

camelid commented on text/000-rustdoc-cfgs-handling.md:
```diff
@@ -0,0 +1,355 @@
+
+### `#[doc(cfg(...))]`
+
+This attribute provides a standardized format to override `#[cfg()]` attributes to document conditionally available items. Example:
```
The phrasing of "override" makes it sound like the _opposite_ of the merging that actually happens between `doc(cfg)` and `doc(auto_cfg)`.

GuillaumeGomez commented on text/000-rustdoc-cfgs-handling.md:
````diff
@@ -0,0 +1,342 @@
+#[doc(auto_cfg(show()))]
+```
+
+But you cannot write:
+
+```rust
+#[doc(auto_cfg(show(not(unix))))]
+```
````
It hides `unix`, `not(unix)`, `any(unix)`, `unix` in `#[cfg(any(unix, windows))]`. It's based on mentions.

camelid replied:
Can you clarify this in the text?

GuillaumeGomez replied:
Sure.

GuillaumeGomez replied:
Done.

GuillaumeGomez commented on text/000-rustdoc-cfgs-handling.md:
````diff
@@ -0,0 +1,342 @@
+Using this attribute will re-enable `auto_cfg` if it was disabled at this location:
+
+```rust
+#[doc(auto_cfg = false)] // Disabling `auto_cfg`
+#[doc(auto_cfg(hide(unix)))] // `auto_cfg` is re-enabled.
+pub fn foo() {}
````
What do you mean?

camelid replied:
Using show and hide for the same cfg on the same item makes no sense, so the RFC stipulates that it should emit an error. IMO disabling auto_cfg on an item but then using `auto_cfg(hide(...))` on the same item _also_ makes no sense and thus should also emit an error. However, you've said that it will instead re-enable auto_cfg. I don't see any reason that that's a desirable behavior.

GuillaumeGomez replied:
I don't see a use case either so we'd better emit an error in this case, indeed. I updated the RFC to change this behaviour.

GuillaumeGomez commented:
Thanks everyone! I'll work on the implementation in the coming days.

nazar-pc commented:
Rendered link should be updated to https://github.com/rust-lang/rfcs/blob/master/text/3631-rustdoc-cfgs-handling.md

jhpratt commented:
Updated
